pkg/services/ngalert/state/historian/loki_http.go

```diff
@@ -11,17 +11,46 @@ import (
 	"time"
 
 	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/setting"
 )
 
 const defaultClientTimeout = 30 * time.Second
 
 type LokiConfig struct {
-	Url               *url.URL
+	ReadPathURL       *url.URL
+	WritePathURL      *url.URL
 	BasicAuthUser     string
 	BasicAuthPassword string
 	TenantID          string
 }
 
+func NewLokiConfig(cfg setting.UnifiedAlertingStateHistorySettings) (LokiConfig, error) {
```

Comment:
Is there any reason why we return the default value instead of a nil? I'm not hugely opinionated either way, but I think this way is perhaps less common in Grafana?

```suggestion
func NewLokiConfig(cfg setting.UnifiedAlertingStateHistorySettings) (*LokiConfig, error) {
```

Reply:
I'm opinionated, and we should stop using Pointer Types everywhere.

Reply:
I've seen this pattern be used for structs in a number of places - the idiom is usually to return the default of the type, not necessarily nil (which only happens to be nil, for pointer types).

Specifically, I chose this because it doesn't make sense for anything to accept a nil LokiConfig anywhere, so I didn't want to introduce nullability into the type when constructing it.

Reply:
The reason I ask is `(*Type, error)` is occurs frequently in the stdlib, `(Type, error)` not so much. For example, `http.NewRequest` and `url.Parse`.

Reply:
The reason I ask is `(*Type, error)` occurs frequently in the stdlib, `(Type, error)` not so much. For example, `http.NewRequest` and `url.Parse`.